### PR TITLE
[SC-26] CFN custom resource for SC actions

### PR DESCRIPTION
**Pull request by zaro0508**
This implements a custom resource to allow us to automate
the creation/deletion/update of service catalog actions using
cloudformation. In addition it also allows us to use cloudformation
to associate SC actions to SC products.

### CODE REVIEW

File: lambdas/cfn-sc-actions-provider/sc-actions-provider/app.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+except Exception as e:
+    helper.init_failure(e)
+
+def get_parameters(event):
+    aws_account_id = event['StackId'].split(':')[4]
+    name = event['ResourceProperties']['Name']
+    ssm_doc_name = event['ResourceProperties']['SsmDocName']
+    ssm_doc_version = event['ResourceProperties']['SsmDocVersion']
+    assume_role = event['ResourceProperties']['AssumeRole']
+    return aws_account_id, name, ssm_doc_name, ssm_doc_version, assume_role
```
[the remark this answers is not in the document]
Comment (zaro0508) [reply lambdas/cfn-sc-actions-provider/sc-actions-provider/app.py:23]: good catch,  crhelper does the`ValueError` validation for us and will automatically return an error to cloudformation with an appropriate message.  In fact I can probably just remove the error handling in the other methods as well.

File: lambdas/cfn-sc-actions-provider/template.yaml [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        Version: '2012-10-17'
+        Statement:
+          - Effect: Allow
+            Action:
+              - logs:CreateLogGroup
+              - logs:CreateLogStream
+              - logs:PutLogEvents
+            Resource: "*"
+
+  ProviderPolicy:
```
Comment (zaro0508) [inline lambdas/cfn-sc-actions-provider/template.yaml:50]: not used right now, this was an attempt to define  a least privileged policy for this lambda however this policy does not work.

File: lambdas/cfn-sc-actions-provider/template.yaml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+              - "ssm:DescribeDocument"
+              - "ssm:GetAutomationExecution"
+              - "ssm:StartAutomationExecution"
+              - "ssm:StopAutomationExecution"
+              - "cloudformation:ListStackResources"
+              - "ec2:DescribeInstanceStatus"
+              - "ec2:StartInstances"
+              - "ec2:StopInstances"
+            Resource: "*"
+#      Policies:
```
Comment (zaro0508) [inline lambdas/cfn-sc-actions-provider/template.yaml:69]: need to remove

File: lambdas/cfn-sc-actions-provider/README.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+```shell script
+sam build --build-dir lambdas/build/cfn-sc-actions-provider --base-dir lambdas/cfn-sc-actions-provider --template lambdas/cfn-sc-actions-provider/template.yaml
+```
+
+## Deploy Lambda to S3
+
+```shell script
+sam package --template-file lambdas/build/cfn-sc-actions-provider/template.yaml --s3-bucket essentials-awss3lambdaartifactsbucket-x29ftznj6pqw --output-template-file templates/cfn-sc-actions-provider.yaml
+```
```
Comment (tthyer) [inline lambdas/cfn-sc-actions-provider/README.md:20]: This is missing the step where the templates are actually copied to S3

File: lambdas/cfn-sc-actions-provider/README.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+template_path: sc-action.yaml
+stack_name: sc-restart-instance-action
+stack_tags:
+  Department: "Platform"
+  Project: "Infrastructure"
+  OwnerEmail: "joe.smith@sagebase.org"
+parameters:
+  Name: "RestartEC2Instance"
+  SsmDocName: "AWS-RestartEC2Instance"
+  SsmDocVersion: "1"
```
Comment (tthyer) [inline lambdas/cfn-sc-actions-provider/README.md:57]: what is the purpose of this version? When will it change?

Comment (zaro0508) [reply lambdas/cfn-sc-actions-provider/README.md:57]: SSM doc can have multiple versions, this is the version number.  check the `events/actions.txt` file for more info and here's the docs on ssm automation docs feature: https://docs.aws.amazon.com/systems-manager/latest/userguide/automation-documents-reference-details.html

File: lambdas/cfn-sc-actions-provider/README.md (hunk 4 above)
Comment (zaro0508) [reply lambdas/cfn-sc-actions-provider/README.md:20]: are you asking a question or making a suggestion?

**Comment by zaro0508** [comment]
Added `events` folder with json events to help you understand.

File: lambdas/cfn-sc-actions-provider/README.md (hunk 4 above)
Comment (tthyer) [reply lambdas/cfn-sc-actions-provider/README.md:20]: making a suggestion -- the directions given in the readme don't work with the deploy-templates.sh script

**Pull request merged**